src/modules/src/supervisor.c

```diff
@@ -0,0 +1,208 @@
+}
+
+//
+// We say we are flying if the sum of the ratio of all motors are above
+// a certain threshold.
+//
+static bool isFlyingCheck()
+{
+  int sumRatio = 0;
+  for (int i = 0; i < NBR_OF_MOTORS; ++i) {
+    sumRatio += motorsGetRatio(i);
+  }
+
+  return sumRatio > SUPERVISOR_FLIGHT_THRESHOLD;
+}
+
+//
+// We say we are tumbled when the accelerometer reports negative values.
+//
+// Once a tumbled situation is identified, we can use this for instance to cut
+// the thrust to the motors, avoiding the Crazyflie from running propellers at
+// significant thrust when accidentally crashing into walls or the ground.
+//
+static bool isTumbledCheck(const sensorData_t *data)
+{
+  const float tolerance = -0.5;
+  static uint32_t hysteresis = 0;
+  //
+  // We need a SUPERVISOR_HYSTERESIS_THRESHOLD amount of readings that indicate
+  // that we are tumbled before we act on it. This is to reduce false positives.
+  //
+  if (data->acc.z <= tolerance) {
+    hysteresis++;
+    if (hysteresis > SUPERVISOR_HYSTERESIS_THRESHOLD) {
+      return true;
+    }
+  } else {
+    hysteresis = 0;
+  }
+
+  return false;
+}
+
+//
+// We detect a free fall by first checking that the total acceleration is close
+// to zero. Then we check the acceleration for z, if it is close to -1 then we
+// can assume a free fall.
+//
+static bool isFreeFallCheck(const sensorData_t *data)
+{
+  const double tolerance = 0.1;
+  static uint32_t hysteresis = 0;
+
+  float totalAcceleration = (data->acc.x * data->acc.x) +
+                            (data->acc.y * data->acc.y) +
+                            (data->acc.z * data->acc.z);
+
+  if (fabs(totalAcceleration) > tolerance) {
+    hysteresis = 0;
+    return false;
+  }
+
+  if (fabs(data->acc.z + 1) <= tolerance) {
+    hysteresis++;
+    if (hysteresis > SUPERVISOR_HYSTERESIS_THRESHOLD) {
+      return true;
+    }
+  } else {
+    hysteresis = 0;
+  }
+
+  return false;
+}
+
+//
+// We detect a Crezyflue at rest by first checking that there is near zero
+// horizontal acceleration. Then we check the acceleration for z, if that is
+// close to one 1 then we can assume that the drone is atRest.We use an order
+// of magnitude larger hysteresis for this check to avoid false positives.
+//
+static bool isAtRestCheck(const sensorData_t *data)
+{
+  const double tolerance = 0.05;
+  static uint32_t hysteresis = 0;
+
+  if (fabs(data->acc.x) > tolerance || fabs(data->acc.y) > tolerance) {
+    hysteresis = 0;
+    return false;
+  }
+
+  if (fabs(data->acc.z - 1) <= tolerance) {
+    hysteresis++;
+    if (hysteresis > SUPERVISOR_HYSTERESIS_THRESHOLD * 10) {
+      return true;
+    }
+  } else {
+    hysteresis = 0;
+  }
+
+  return false;
+}
+
+void supervisorUpdate(const sensorData_t *data)
+{
+  isFlying = isFlyingCheck();
+
+  isFreeFall = isFreeFallCheck(data);
+
+  isAtRest = isAtRestCheck(data);
+
+  isTumbled = isTumbledCheck(data);
+  if (isTumbled && isFlying) {
+    stabilizerSetEmergencyStop();
```

Comment:
I think the emergency stop functionality would also be better part of the supervisor? Could be done in a separate change, though.

Reply:
You mean moving the `stabilizerSetEmergencyStop()` function into superviser.c?

Reply:
Yes, that's what I meant. From a logical functionality perspective this seems to be supervisor-worthy; this should not block this PR, though.